nlp_utilities.jl

```diff
@@ -0,0 +1,408 @@
+
+    # Function Derivatives
+    hessian, jacobian = compute_optimal_hess_jac(evaluator, cons, all_vars)
+
+    # Hessian of the lagrangian wrt the primal variables
+    W = zeros(num_vars + num_ineq, num_vars + num_ineq)
+    W[1:num_vars, 1:num_vars] = hessian[1:num_vars, 1:num_vars]
+    # Jacobian of the constraints wrt the primal variables
+    A = zeros(num_cons, num_vars + num_ineq)
+    A[:, 1:num_vars] = jacobian[:, 1:num_vars]
+    for (i,j) in enumerate(ineq_locations)
+        A[j, num_vars+i] = -1
+    end
+    # Partial second derivative of the lagrangian wrt primal solution and parameters
+    ∇ₓₚL = zeros(num_vars + num_ineq, num_parms)
+    ∇ₓₚL[1:num_vars, :] = hessian[1:num_vars, num_vars+1:end]
+    # Partial derivative of the equality constraintswith wrt parameters
+    ∇ₚC = jacobian[:, num_vars+1:end]
+
+    # M matrix
+    # M = [
+    #     [W A' -I I];
+    #     [A 0 0 0];
+    #     [V_L 0 (X - X_L) 0]
+    #     [V_U 0 0 0 (X_U - X)]
+    # ]
+    len_w = num_vars + num_ineq
+    M = zeros(len_w + num_cons + num_low + num_up, len_w + num_cons + num_low + num_up)
+
+    M[1:len_w, 1:len_w] = W
+    M[1:len_w, len_w + 1 : len_w + num_cons] = A'
+    M[len_w+1:len_w+num_cons, 1:len_w] = A
+    M[1:len_w, len_w+num_cons+1:len_w+num_cons+num_low] = I_L
+    M[len_w+num_cons+1:len_w+num_cons+num_low, 1:len_w] = V_L
+    M[len_w+num_cons+1:len_w+num_cons+num_low, len_w+num_cons+1:len_w+num_cons+num_low] = X_lb
+    M[len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up, 1:len_w] = V_U
+    M[len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up, len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up] = X_ub
+    M[1:len_w, len_w+num_cons+num_low+1:end] = I_U
+
+    # N matrix
+    N = [∇ₓₚL ; ∇ₚC; zeros(num_low + num_up, num_parms)]
+
+    return M, N
+end
+
+"""
+    compute_derivatives_no_relax(evaluator::MOI.Nonlinear.Evaluator, cons::Vector{ConstraintRef},
+        primal_vars::Vector{VariableRef}, params::Vector{VariableRef},
+        _X::Vector{T}, _V_L::Vector{T}, _X_L::Vector{T}, _V_U::Vector{T}, _X_U::Vector{T}, ineq_locations::Vector{Z},
+        has_up::Vector{Z}, has_low::Vector{Z}
+    ) where {T<:Real, Z<:Integer}
+
+Compute the derivatives of the solution w.r.t. the parameters without accounting for active set changes.
+"""
+function compute_derivatives_no_relax(evaluator::MOI.Nonlinear.Evaluator, cons::Vector{ConstraintRef},
+    primal_vars::Vector{VariableRef}, params::Vector{VariableRef}, 
+    _X::Vector{T}, _V_L::Vector{T}, _X_L::Vector{T}, _V_U::Vector{T}, _X_U::Vector{T}, ineq_locations::Vector{Z},
+    has_up::Vector{Z}, has_low::Vector{Z}
+) where {T<:Real, Z<:Integer}
+    num_bounds = length(has_up) + length(has_low)
+    M, N = build_M_N(evaluator, cons, primal_vars, params, _X, _V_L, _X_L, _V_U, _X_U, ineq_locations, has_up, has_low)
+
+    # Sesitivity of the solution (primal-dual_constraints-dual_bounds) w.r.t. the parameters
+    K = qr(M) # Factorization
+    ∂s = - (K \ N) # Sensitivity
```

Comment:
I am trying to pass the code to only use sparse matrices but getting some problems on the backsolve:

```
Got exception outside of a @test
  MethodError: no method matching ldiv!(::SparseArrays.UMFPACK.UmfpackLU{Float64, Int64}, ::SparseMatrixCSC{Float64, Int64})
  
  Closest candidates are:
    ldiv!(::Bidiagonal, ::AbstractVecOrMat)
     @ LinearAlgebra ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/LinearAlgebra/src/bidiag.jl:721
    ldiv!(::LQ, ::AbstractVecOrMat)
     @ LinearAlgebra ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/LinearAlgebra/src/lq.jl:185
    ldiv!(::Diagonal{T, V} where V<:AbstractVector{T}, ::SparseArrays.AbstractSparseMatrixCSC{T}) where T
     @ SparseArrays ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/SparseArrays/src/linalg.jl:1016
    ...
  
  Stacktrace:
    [1] \(F::SparseArrays.UMFPACK.UmfpackLU{Float64, Int64}, B::SparseMatrixCSC{Float64, Int64})
      @ LinearAlgebra ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/LinearAlgebra/src/factorization.jl:143
    [2] compute_derivatives_no_relax(evaluator::MathOptInterface.Nonlinear.Evaluator{MathOptInterface.Nonlinear.ReverseAD.NLPEvaluator}, cons::Vector{ConstraintRef}, primal_vars::Vector{VariableRef}, params::Vector{VariableRef}, _X::Vector{Float64}, _V_L::Vector{Float64}, _X_L::Vector{Float64}, _V_U::Vector{Float64}, _X_U::Vector{Float64}, ineq_locations::Vector{Int64}, has_up::Vector{Int64}, has_low::Vector{Int64})
      @ Main ~/workspace/DiffOpt.jl/nlp_utilities.jl:307 
```

Perhaps this might be able to help but unsure if including it is worth it: https://github.com/JuliaSmoothOptimizers/BasicLU.jl?tab=readme-ov-file

Reply:
Trying the BasicLU package now but getting a few numerical errors

Reply:
If this is something we might want to explore, I can branch this branch and keep trying to make it work with BasicLU. I am open to suggestions on how to deal with the sparse matrix backsolve.

Reply:
use `ldiv!` directly.